src/miner.cpp

```diff
@@ -473,6 +515,141 @@ void BlockAssembler::addScoreTxs(std::vector<const CTxMemPoolEntry *> *vtxe)
+// could for instance end up including a lower fee transaction as part of an ancestor group when
+// in fact it would be better, in terms of fees, to include some other single transaction. This
+// would result in slightly less fees (perhaps a few hundred satoshis) rewarded to the miner. However,
+// this situation is not likely to be seen for two reasons. One, long unconfirmed chains are typically
+// having transactions with all the same fees and Two, the typical child pays for parent scenario has only
+// two transactions with the child having the higher fee. And neither of these two types of packages could
+// cause any loss of fees with this mining algorithm, when the block is nearly full.
+void BlockAssembler::addPackageTxs(std::vector<const CTxMemPoolEntry *> *vtxe, bool fCanonical)
+{
+    AssertLockHeld(mempool.cs);
+
+    CTxMemPool::indexed_transaction_set::index<ancestor_score>::type::iterator mi =
+        mempool.mapTx.get<ancestor_score>().begin();
+    CTxMemPool::txiter iter;
+
+    uint64_t nPackageFailures = 0;
+    while (mi != mempool.mapTx.get<ancestor_score>().end())
+    {
+        iter = mempool.mapTx.project<0>(mi);
+        ++mi;
+
+        uint64_t packageSize = iter->GetSizeWithAncestors();
+        CAmount packageFees = iter->GetModFeesWithAncestors();
+        unsigned int packageSigOps = iter->GetSigOpCountWithAncestors();
+
+        // Skip txns we know are in the block
+        if (SkipMapTxEntry(iter))
+        {
+            continue;
+        }
+
+        // Get any unconfirmed ancestors of this txn
+        CTxMemPool::setEntries ancestors;
+        uint64_t nNoLimit = std::numeric_limits<uint64_t>::max();
+        std::string dummy;
+        mempool._CalculateMemPoolAncestors(
+            *iter, ancestors, nNoLimit, nNoLimit, nNoLimit, nNoLimit, dummy, &inBlock, false);
+
+        // Include in the package the current txn we're working with
+        ancestors.insert(iter);
+
+        // Recalculate sigops and package size, only if there were txns already in the block for
+        // this set of ancestors
+        if (iter->GetCountWithAncestors() > ancestors.size())
+        {
+            packageSize = 0;
+            packageSigOps = 0;
+            for (auto &it : ancestors)
+            {
+                packageSize += it->GetTxSize();
+                packageSigOps += it->GetSigOpCount();
+            }
+        }
+        if (packageFees < ::minRelayTxFee.GetFee(packageSize) && nBlockSize >= nBlockMinSize)
+        {
+            // Everything else we might consider has a lower fee rate so no need to continue
+            return;
+        }
+
+        // Test if package fits in the block
+        if (nBlockSize + packageSize > nBlockMaxSize)
+        {
+            if (nBlockSize > nBlockMaxSize * .50)
+            {
+                nPackageFailures++;
+            }
+
+            // If we keep failing then the block must be almost full so bail out here.
+            if (nPackageFailures >= 5)
```

Comment:
Let's make 5 a symbolic constant, I can imagine that we might want to tune this.

Reply:
done